tests/spec/README.md

```diff
@@ -0,0 +1,112 @@
+# Spec tests
+
+These tests specify the expected behaviour of ehrQL, and by extension, the QM.
+
+They are also used to generate documentation.
+
+
+## Table of Contents
+
+`toc.py` defines the Table of Contents for the documentation generated by tests. A new spec test
+subdirectory or file must be defined here in order for it to appear in the docs.
+
+## Folder structure
+
+Within the spec directory, each test directory is considered a chapter in the docs, each file
+is considered to be a section, and each test is a paragraph.
+
+e.g.
+```
+tests
+    ├── spec
+    │   ├── aggregate_frame                       <-- chapter
+    │   │   ├── __init__.py                       <-- contains chapter title
+    │   │   ├── test_count_for_patient.py         <-- section
+    │   │   └── test_exists_for_patient.py        <-- section
+```
+
+### Titles
+Chapter titles are specified with a `title` attribute the test directory's `__init__.py`.
+Section titles are specified with a `title` attribute in each test file.
+Paragraph titles are extracted from the test names.
+
+E.g. `test_count_for_patient.py` contains 2 tests:
+- `test_count_for_patient_on_event_frame()`
+- `test_count_for_patient_on_patient_frame()`
+
+Assuming this is included in `toc.py`, the following structure will be generated in the docs:
+
+1. Aggregating event and patient frames  # from `title` in `aggregate_frame/__init__.py`
+1.1 Counting the rows for each patient  # from `title` in `aggregate_frame/test_count_for_patient.py`
+1.1.1 Count for patient on event frame  # from test name
+1.1.2 Count for patient on patient frame  # from test name
+
+
+## Test structure
+
+Spec tests follow a standardised structure in order to allow documentation generation.  Each
+spec test uses the `spec_test` fixture, which has the following components:
+
+```
+spec_test(
+    table_data,                          # defines the tables in the test database
+    e.take(e.b1).i1.sum_for_patient(),   # the ehrQL code being tested
+    {                                    # expected results
+        1: (101 + 102),
+        2: 201,
+    },
+)
+```
+
+### Defining table data
+Each test must define a `table_data` dict, with keys `e` (event level data, with multiple rows per patient) and/or `p` (patient level data, with one row per patient)
+
+This sets up the test data.  The first column is always patient ID.  Column names indicate the type of the column in the test database.
+
+E.g. The following table data dict sets up an event table with an integer column `i1` and a
+boolean column `b1`, with 2 rows for each of 2 patients, and a patient table with a date column
```

Comment:
question: Can you explain what's being counted here, and how, for the reader? It's a little ambiguous to me as written.

To my naive eyes, `b1` in table `e` looks like it has:

* three rows for patient ID `1`
* one non-null row and one null row for patient ID `2`, so two rows total

I can't quite see how `b1` has:

> 2 rows for each of 2 patients

so either it might benefit from explanation, or I'm misreading something here

Reply:
I think I added an additional row for patient 1 and forgot to update the description